wgengine/router/router_linux.go

```diff
@@ -442,95 +552,124 @@ func (r *linuxRouter) delBypassRule() error {
 	if !bytes.Contains(out, []byte(" fwmark "+tailscaleBypassMark)) {
 		return fmt.Errorf("ip rule 10000 doesn't look like a Tailscale policy rule: %q", string(out))
 	}
-	return cmd("ip", "rule", "del", "priority", "10000")
+	return r.cmd.run("ip", "rule", "del", "priority", "10000")
 }
 
-// deleteLegacyNetfilter removes the netfilter rules installed by
-// older versions of Tailscale, if they exist.
-func (r *linuxRouter) deleteLegacyNetfilter() error {
-	del := func(table, chain string, args ...string) error {
-		exists, err := r.ipt4.Exists(table, chain, args...)
+// addNetfilterBase adds custom Tailscale chains to netfilter, along
+// with some basic processing rules to be supplemented by later calls
+// to other helpers.
+func (r *linuxRouter) addNetfilterBase() error {
+	create := func(table, chain string) error {
+		chains, err := r.ipt4.ListChains(table)
 		if err != nil {
-			return fmt.Errorf("checking for %v in %s/%s: %v", args, table, chain, err)
+			return fmt.Errorf("listing iptables chains: %v", err)
 		}
-		if exists {
-			if err := r.ipt4.Delete(table, chain, args...); err != nil {
-				return fmt.Errorf("deleting %v in %s/%s: %v", args, table, chain, err)
+		found := false
+		for _, ch := range chains {
+			if ch == chain {
+				found = true
+				break
 			}
 		}
+		if found {
+			err = r.ipt4.ClearChain(table, chain)
+		} else {
+			err = r.ipt4.NewChain(table, chain)
+		}
+		if err != nil {
+			return fmt.Errorf("setting up %s/%s: %v", table, chain, err)
+		}
 		return nil
 	}
-
-	if err := del("filter", "FORWARD", "-m", "comment", "--comment", "tailscale", "-i", r.tunname, "-j", "ACCEPT"); err != nil {
+	if err := create("filter", "ts-input"); err != nil {
 		return err
 	}
-	if err := del("nat", "POSTROUTING", "-m", "comment", "--comment", "tailscale", "-o", "eth0", "-j", "MASQUERADE"); err != nil {
+	if err := create("filter", "ts-forward"); err != nil {
+		return err
+	}
+	if err := create("nat", "ts-postrouting"); err != nil {
 		return err
 	}
 
+	// Only allow CGNAT range traffic to come from tailscale0. There
+	// is an exception carved out for ranges used by ChromeOS, for
+	// which we fall out of the Tailscale chain.
+	//
+	// Note, this will definitely break nodes that end up using the
+	// CGNAT range for other purposes :(.
```

Comment:
Can we detect if they are and warn appropriately? (E.g. look to see if a bad ISP has assigned a cgnat address to any non-tailscale interfaces?)

Reply:
Filed https://github.com/tailscale/tailscale/issues/389 for that.